rust/src/http2/http2.rs

```diff
@@ -0,0 +1,1158 @@
+                            {
+                                self.set_event(HTTP2Event::InvalidHeader);
+                            }
+                        }
+                        if hrem.len() > 0 {
+                            SCLogDebug!("Remaining data for HTTP2 headers");
+                            self.set_event(HTTP2Event::ExtraHeaderData);
+                        }
+                        return HTTP2FrameTypeData::HEADERS(hs);
+                    }
+                    Err(nom::Err::Incomplete(_)) => {
+                        if complete {
+                            self.set_event(HTTP2Event::InvalidFrameData);
+                            return HTTP2FrameTypeData::UNHANDLED(HTTP2FrameUnhandled {
+                                reason: HTTP2FrameUnhandledReason::ParsingError,
+                            });
+                        } else {
+                            return HTTP2FrameTypeData::UNHANDLED(HTTP2FrameUnhandled {
+                                reason: HTTP2FrameUnhandledReason::TooLong,
+                            });
+                        }
+                    }
+                    Err(_) => {
+                        self.set_event(HTTP2Event::InvalidFrameData);
+                        return HTTP2FrameTypeData::UNHANDLED(HTTP2FrameUnhandled {
+                            reason: HTTP2FrameUnhandledReason::ParsingError,
+                        });
+                    }
+                }
+            }
+            Some(parser::HTTP2FrameType::PING) => {
+                return HTTP2FrameTypeData::PING;
+            }
+            _ => {
+                return HTTP2FrameTypeData::UNHANDLED(HTTP2FrameUnhandled {
+                    reason: HTTP2FrameUnhandledReason::UnknownType,
+                });
+            }
+        }
+    }
+
+    fn stream_data(&mut self, dir: u8, input: &[u8], over: bool, txid: u64) {
+        match unsafe { SURICATA_HTTP2_FILE_CONFIG } {
+            Some(sfcm) => {
+                for tx in &mut self.transactions {
+                    if tx.tx_id == txid {
+                        let xid: u32 = tx.tx_id as u32;
+                        let (files, flags) = self.files.get(dir);
+                        tx.ft.tx_id = tx.tx_id - 1;
+                        tx.ft.new_chunk(
+                            sfcm,
+                            files,
+                            flags,
+                            b"",
+                            input,
+                            tx.ft.tracked, //offset = append
+                            input.len() as u32,
+                            0,
+                            over,
+                            &xid,
+                        );
+                        break;
+                    }
+                }
+            }
+            None => panic!("BUG"),
+        }
+    }
+
+    fn parse_ts(&mut self, mut input: &[u8]) -> AppLayerResult {
```

Comment:
functions like this are very long. I think a goal should be to reduce the size, ideally to always fit completely in a single (reasonably sized) screen.

Reply:
ok done, and factorized some code with `parse_tc`